op-node/rollup/sync/start.go

```diff
@@ -41,163 +28,197 @@ import (
+			// Check L2 traversal against L2 Genesis data, to make sure the engine is on the correct chain, instead of attempting sync with different L2 destination.
+			if n.Hash != cfg.Genesis.L2.Hash {
+				return nil, fmt.Errorf("%w L2: genesis: %s, got %s", WrongChainErr, cfg.Genesis.L2, n)
+			}
+			// Check L1 comparison against L1 Genesis data, to make sure the L1 data is from the correct chain, instead of attempting sync with different L1 source.
+			if !ahead && l1Block.Hash != cfg.Genesis.L1.Hash {
+				return nil, fmt.Errorf("%w L1: genesis: %s, got %s", WrongChainErr, cfg.Genesis.L1, l1Block)
+			}
 		}
-
-		// Pull L2 parent for next iteration
-		n, err = l2.L2BlockRefByHash(ctx, n.ParentHash)
-		if err != nil {
-			return eth.L2BlockRef{}, eth.L2BlockRef{},
-				fmt.Errorf("failed to fetch L2 block by hash %v: %w", n.ParentHash, err)
+		// Check L2 traversal against finalized data
+		if (n.Number == result.Finalized.Number) && (n.Hash != result.Finalized.Hash) {
+			return nil, fmt.Errorf("%w: finalized %s, got: %s", ReorgFinalizedErr, result.Finalized, n)
 		}
-
-		reorgDepth++
-		if reorgDepth >= MaxReorgDepth {
+		// Check we are not reorging L2 incredibly deep
+		if n.Number+MaxReorgDepth < prevUnsafe.Number {
 			// If the reorg depth is too large, something is fishy.
 			// This can legitimately happen if L1 goes down for a while. But in that case,
 			// restarting the L2 node with a bigger configured MaxReorgDepth is an acceptable
 			// stopgap solution.
-			// Currently this can also happen if the L2 node is down for a while, but in the future
-			// state sync should prevent this issue.
-			return eth.L2BlockRef{}, eth.L2BlockRef{}, TooDeepReorgErr
+			return nil, fmt.Errorf("%w: traversed back to L2 block %s, but too deep compared to previous unsafe block %s", TooDeepReorgErr, n, prevUnsafe)
 		}
-	}
 
-	// Loop 2. Walk from the L1 origin of the `n` block (*) back to the L1 block that starts the
-	// sequencing window ending at that block. Instead of iterating on L1 blocks, we actually
-	// iterate on L2 blocks, because we want to find the safe L2 head, i.e. the highest L2 block
-	// whose L1 origin is the start of the sequencing window.
+		// If we don't have a usable unsafe head, then set it
+		if result.Unsafe == (eth.L2BlockRef{}) {
+			result.Unsafe = n
+		}
 
-	// (*) `n` being at this stage the highest L2 block whose L1 origin is canonical.
+		if ahead {
+			// keep the unsafe head if we can't tell if its L1 origin is canonical or not yet.
+		} else if l1Block.Hash == n.L1Origin.Hash {
+			// if L2 matches canonical chain, even if unsafe,
+			// then we can start finding a span of L1 blocks to cover the sequence window,
+			// which may help avoid rewinding the existing safe head unnecessarily.
+			if highestL2WithCanonicalL1Origin == (eth.L2BlockRef{}) {
+				highestL2WithCanonicalL1Origin = n
+			}
+		} else {
+			// L1 origin nor ahead of L1 head nor canonical, discard previous candidate and keep looking.
```

Comment:
```suggestion
			// L1 origin not ahead of L1 head nor canonical, discard previous candidate and keep looking.
```

Reply:
nice catch, copied the comment over and didn't see the typo.